riptide-compression/src/main/java/org/zalando/riptide/compression/RequestCompressionPlugin.java

```diff
@@ -0,0 +1,88 @@
+package org.zalando.riptide.compression;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apiguardian.api.API;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpOutputMessage;
+import org.springframework.http.StreamingHttpOutputMessage;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestArguments.Entity;
+import org.zalando.riptide.RequestExecution;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.zip.GZIPOutputStream;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+import static org.springframework.http.HttpHeaders.CONTENT_ENCODING;
+import static org.springframework.http.HttpHeaders.TRANSFER_ENCODING;
+
+@API(status = EXPERIMENTAL)
+public final class RequestCompressionPlugin implements Plugin {
+
+    private final Compression compression;
+
+    public RequestCompressionPlugin(final Compression compression) {
+        this.compression = compression;
+    }
+
+    public RequestCompressionPlugin() {
+        this(Compression.of("gzip", GZIPOutputStream::new));
+    }
+
+    @Override
+    public RequestExecution aroundNetwork(final RequestExecution execution) {
+        return arguments -> {
+            final Entity entity = arguments.getEntity();
+
+            if (entity.isEmpty()) {
+                return execution.execute(arguments);
+            }
+
+            return execution.execute(
+                    arguments.withEntity(new CompressingEntity(compression, entity)));
+        };
+    }
+
+    @AllArgsConstructor
+    private static class CompressingEntity implements Entity {
+
+        private final Compression compression;
+        private final Entity entity;
+
+        @Override
+        public void writeTo(final HttpOutputMessage message) throws IOException {
+            update(message.getHeaders());
+
+            if (message instanceof StreamingHttpOutputMessage) {
+                final StreamingHttpOutputMessage streaming = (StreamingHttpOutputMessage) message;
+                streaming.setBody(stream ->
+                        writeToCompressing(new DelegatingHttpOutputMessage(message.getHeaders(), stream)));
+            } else {
+                writeToCompressing(message);
+            }
+        }
+
+        private void writeToCompressing(HttpOutputMessage message) throws IOException {
+            try (final WrappingHttpOutputMessage compressingMessage =
+                         new WrappingHttpOutputMessage(message, compression.getOutputStreamWrapper())) {
+                entity.writeTo(compressingMessage);
+            }
+        }
+
+        private void update(final HttpHeaders headers) {
+            headers.set(CONTENT_ENCODING, compression.getContentEncoding());
```

Comment:
So effectively if somebody will try to configure multiple EncodingPlugins (for what ever reason... send request as `tar.gz`) you'll overwrite the header with the last encoding specified

Reply:
Yes currently not supported, see discussion https://github.com/zalando/riptide/pull/822#discussion_r346167494

Reply:
Was not clear from that discussion, neither I saw in docs something like "Don't use multiple instances for the same plugin with different configs".

Reply:
This draw back may be documented yes - is there a way to ensure only one plugin of a type is registered at a time? Feels weird though.

Reply:
I think there can be a way (i.e. spring will reject multiple auto wiring candidates unless you mark one as primary), but this will require an reviewing of the whole plugin setup to distinguish what plugins need to be singleton and which will support chaining

Reply:
One pattern that I used successfully is to make all plugins behave nice if they can detect that someone already did something, e.g. the `AuthorizationPlugin` doesn't do anything if the `Authorization` header is already present, for whatever reason.

I'd do the same based on `Content-Encoding`.

Reply:
Adapted.